lite/src/sequential.rs

```diff
@@ -0,0 +1,102 @@
+use crate::types::*;
+
+/// SequentialVerifier verifies all headers in order,
+/// where +2/3 of the correct validator set must have signed the header.
+struct SequentialVerifier<H, V>
+where
+    H: Header,
+    V: Validators,
+{
+    trusting_period: Time,
+    state: TrustedState<H, V>,
+}
+
+impl<H, V> SequentialVerifier<H, V>
+where
+    H: Header,
+    V: Validators,
+{
+    /// Returns the time after which the current state expires
+    /// and the verifier must be reset subjectively.
+    fn expires(&self) -> Time {
+        self.state.header.bft_time() + self.trusting_period
+    }
+
+    /// Verify takes a header, a commit for the header,
+    /// and the next validator set referenced by the header.
+    /// Without knowing this next validator set, we can't
+    /// really verify the next header, so we make verifying
+    /// the current header conditional on receiving that validator set.
+    /// We could break this up into two steps, one where we verify the header,
+    /// and one where we receive and validate the next validators,
+    /// but for now we combine them. Returns an error if verification fails.
+    fn verify<C>(&self, now: Time, header: H, commit: C, next_validators: V) -> Result<(), Error>
+    where
+        C: Commit,
+    {
+        // ensure the state is not expired.
+        if self.expires() < now {
+            return Err(Error::Expired);
+        }
+
+        // ensure the height is strictly sequential.
+        if header.height() != self.state.header.height() + 1 {
+            return Err(Error::NonSequentialHeight);
+        }
+
+        // ensure the validators in the header matches what we expect from our state.
+        if header.validators_hash() != self.state.next_validators.hash() {
+            return Err(Error::InvalidValidators);
+        }
+
+        // ensure the next validators in the header matches what was supplied.
+        if header.next_validators_hash() != next_validators.hash() {
+            return Err(Error::InvalidNextValidators);
+        }
+
+        // ensure the commit matches the header.
+        if header.hash() != commit.header_hash() {
+            return Err(Error::InvalidCommitValue);
+        }
+
+        // ensure that +2/3 validators signed correctly
+        self.verify_commit(commit)
+    }
+
+    /// Check that +2/3 of the correct and trusted validator set signed this commit.
+    fn verify_commit<C>(&self, commit: C) -> Result<(), Error>
+    where
+        C: Commit,
+    {
+        let total_power = self.state.next_validators.total_power();
+        let mut signed_power: u64 = 0;
+
+        // NOTE that the vals and commit have a 1-to-1 correspondance.
+        // This means we don't need the validator IDs or to do any lookup,
+        // we can just zip the iterators.
```

Comment:
IMO we should still verify it, i.e. check same number/ length of vals and commits.

Reply:
Done.